### PR TITLE
Avoid boxing in AccessLists

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Maintain two strongly typed `List<T>` to avoid boxing the `UInt256` keys to `object` (some txs have 6k+)
- Clean up some `UInt256 == 0` to `.IsZero`
- Also long comparision to 0 rather than convert to UInt256 and then compare zero
- Use Span when calcuating DataCost

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

**Review by LukaszRozmej — COMMENTED** [review]
Does this preserve the order of access list as before when serializing/deserializing RLP? Different access list order between roundtrip would result in transaction hash mismatch.

File: src/Nethermind/Nethermind.Evm/EvmState.cs [hunk 1]
```diff
@@ -273,7 +273,7 @@ public void InitStacks()
 
         public void WarmUp(AccessList? accessList)
         {
-            if (accessList is not null)
+            if (accessList is not null && !accessList.IsEmpty)
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/EvmState.cs]: Changed

File: src/Nethermind/Nethermind.Evm/Tracing/Proofs/ProofTxTracer.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public override void ReportBlockHash(Hash256 blockHash)
 
         public override void ReportBalanceChange(Address address, UInt256? before, UInt256? after)
         {
-            if (_treatSystemAccountDifferently && Address.SystemUser == address && before is null && after == UInt256.Zero)
+            if (_treatSystemAccountDifferently && Address.SystemUser == address && before is null && (after?.IsZero ?? true))
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Evm/Tracing/Proofs/ProofTxTracer.cs]: Changed

**Comment by benaadams** [comment]
> Does this preserve the order of access list as before when serializing/deserializing RLP?

I assume so as it doing same order as it was before this change

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**